llvm/lib/IR/IRBuilder.cpp

```diff
@@ -422,6 +436,11 @@ CallInst *IRBuilderBase::CreateFAddReduce(Value *Acc, Value *Src) {
   return CreateCall(Decl, Ops);
 }
 
+CallInst *IRBuilderBase::CreateFAddReduce(Value *Acc, Value *Src, Value *EVL,
+                                          Value *Mask) {
+  return getReductionIntrinsic(Intrinsic::vp_reduce_fadd, Acc, Src, Mask, EVL);
```

Comment:
Can we avoid this duplication here by providing a builder that provides an API to create an fadd reduction, which either creates `vector_reduce_fadd` if no mask or EVL has been provided and `vp_reduce_fadd` if mask and EVL has been provided, e.g. by extending `VectorBuilder`?

Then at least the complexity would be unified at a single place.

Reply:
We had couple of reasons to duplicate code and extend IRBuilder:
* Less changes in vectorizer and less painful to do pulldown. If VectorBuilder will be used during in VPReductionRecipe, that's not important reason anymore.
* More concrete expectations and not need to handle all variations:
** EVL != nullptr && Mask != nullptr
** EVL != nulltptr && Mask == nullptr
** EVL == nullptr && Mask == nullptr
** EVL == nullptr && Mask != nullptr
Unless there is a plan to support latter, it doesn't look good to pretend that scenario is supported.
BTW, we should really have `&EVL`

Reply:
172745025d81388ef8104006ffd7e6faa3d1b04a
@fhahn @nikolaypanchenko I have attempted to move the utils function of IRBuilderBase to VectorBuilder. Please review and let me know if this design is appropriate. Thank you.

Reply:
The latest version moves the logic to create vector-predication intrinsics to the vector-intrinsic builder, which addressed my main concern (re-use of logic elsewhere), so that looks good to me for now